storm-core/src/jvm/backtype/storm/blobstore/BlobStore.java

```diff
@@ -0,0 +1,446 @@
+
+  /**
+   * Filters keys based on the KeyFilter
+   * passed as the argument.
+   * @param filter Filter passed as the
+   * @param who Might not want to have the subject as it is not doing anything
+   * @param <R> Type
+   * @return Set of filtered keys
+   */
+  public <R> Set<R> filterAndListKeys(KeyFilter<R> filter, Subject who) {
+    Set<R> ret = new HashSet<R>();
+    Iterator<String> keys = listKeys(who);
+    while (keys.hasNext()) {
+      String key = keys.next();
+      R filtered = filter.filter(key);
+      if (filtered != null) {
+        ret.add(filtered);
+      }
+    }
+    return ret;
+  }
+
+  /**
+   * Validates key checking for potentially harmful patterns
+   * @param key Key for the blob.
+   * @throws AuthorizationException
+   */
+  public static final void validateKey(String key) throws AuthorizationException {
+    if (key == null || key.isEmpty() || "..".equals(key) || ".".equals(key) || !KEY_PATTERN.matcher(key).matches()) {
+      LOG.error("'{}' does not appear to be valid {}", key, KEY_PATTERN);
+      throw new AuthorizationException(key+" does not appear to be a valid blob key");
+    }
+  }
+
+  /**
+   * Wrapper called to create the blob which contains
+   * the byte data
+   * @param key Key for the blob.
+   * @param data Byte data that needs to be uploaded.
+   * @param meta Metadata which contains the acls information
+   * @param who Is the subject creating the blob.
+   * @throws AuthorizationException
+   * @throws KeyAlreadyExistsException
+   * @throws IOException
+   */
+  public void createBlob(String key, byte [] data, SettableBlobMeta meta, Subject who) throws AuthorizationException, KeyAlreadyExistsException, IOException {
+    AtomicOutputStream out = null;
+    try {
+      out = createBlob(key, meta, who);
+      out.write(data);
+      out.close();
+      out = null;
+    } finally {
+      if (out != null) {
+        out.cancel();
+      }
+    }
+  }
+
+  /**
+   * Wrapper called to create the blob which contains
+   * the byte data
+   * @param key Key for the blob.
+   * @param in InputStream from which the data is read to be
+   * written as a part of the blob.
+   * @param meta Metadata which contains the acls information
+   * @param who Is the subject creating the blob.
+   * @throws AuthorizationException
+   * @throws KeyAlreadyExistsException
+   * @throws IOException
+   */
+  public void createBlob(String key, InputStream in, SettableBlobMeta meta, Subject who) throws AuthorizationException, KeyAlreadyExistsException, IOException {
+    AtomicOutputStream out = null;
+    try {
+      out = createBlob(key, meta, who);
+      byte[] buffer = new byte[2048];
+      int len = 0;
+      while ((len = in.read(buffer)) > 0) {
+        out.write(buffer, 0, len);
+      }
+      out.close();
+    } catch (AuthorizationException | IOException | RuntimeException e) {
+      out.cancel();
```

Comment:
If we catch `AuthorizationException` from the above call to `createBlob`, then `out` will be null here.  So we need to check that `out` is not null here before calling `cancel`.